stdlib/src/collections/list.mojo

```diff
@@ -568,3 +572,36 @@ struct List[T: CollectionElement](CollectionElement, Sized, Boolable):
         """
         var ref = Reference(self)
         return _ListIter[T, mutability, self_life, False](len(ref[]), ref)
+
+    @staticmethod
+    fn __str__[U: StringableCollectionElement](self: List[U]) -> String:
+        """Returns a string representation of a `List`.
+
+        Note that since we can't condition methods on a trait yet,
+        the way to call this method is a bit special. Here is an example below:
+
+        ```mojo
+        var my_list = List[Int](1, 2, 3)
+        print(__type_of_(my_list).__str__(my_list))
+        ```
+
+        When the compiler supports conditional methods, then a simple `str(my_list)` will
+        be enough.
+
+        Args:
+            self: The list to represent as a string.
+
+        Parameters:
+            U: The type of the elements in the list. Must implement the
+              traits `Stringable` and `CollectionElement`.
+
+        Returns:
+            A string representation of the list.
+        """
+        # TODO: Use `String.join()` when it's working with List.
+        var result = String("[")
```

Comment:
**Suggestion** Would it make sense to have the result be a `List[UInt8]` and initialize its capacity ahead of time with the pre-computed size?  Then we could fill in the buffer rather than doing some potential reallocations/copies in the `+=` calls.

Reply:
Done. If str(x) is one char, we shouldn't get any allocations.